store/tikv/mock-tikv/mvcc.go

```diff
@@ -0,0 +1,323 @@
+		Key:     e.key,
+		Primary: e.lock.primary,
+		StartTS: e.lock.startTS,
+	}
+}
+
+func (e *mvccEntry) Get(ts uint64) ([]byte, error) {
+	if e.lock != nil {
+		if e.lock.startTS <= ts {
+			return nil, e.lockErr()
+		}
+	}
+	for _, v := range e.values {
+		if v.commitTS <= ts {
+			return v.value, nil
+		}
+	}
+	return nil, nil
+}
+
+func (e *mvccEntry) Prewrite(mutation *kvrpcpb.Mutation, startTS uint64, primary []byte) error {
+	if len(e.values) > 0 {
+		if e.values[0].commitTS >= startTS {
+			return ErrRetryable("write conflict")
+		}
+	}
+	if e.lock != nil {
+		if e.lock.startTS != startTS {
+			return e.lockErr()
+		}
+		return nil
+	}
+	e.lock = &mvccLock{
+		startTS: startTS,
+		primary: primary,
+		value:   mutation.Value,
+		op:      mutation.GetOp(),
+	}
+	return nil
+}
+
+func (e *mvccEntry) checkTxnCommitted(startTS uint64) (uint64, bool) {
+	for _, v := range e.values {
+		if v.startTS == startTS {
+			return v.commitTS, true
+		}
+	}
+	return 0, false
+}
+
+func (e *mvccEntry) Commit(startTS, commitTS uint64) error {
+	if e.lock == nil || e.lock.startTS != startTS {
+		if _, ok := e.checkTxnCommitted(startTS); ok {
+			return nil
+		}
+		return ErrRetryable("txn not found")
+	}
+	if e.lock.op != kvrpcpb.Op_Lock {
+		e.values = append([]mvccValue{{
+			startTS:  startTS,
+			commitTS: commitTS,
+			value:    e.lock.value,
+		}}, e.values...)
+	}
+	e.lock = nil
+	return nil
+}
+
+func (e *mvccEntry) Rollback(startTS uint64) error {
+	if e.lock == nil || e.lock.startTS != startTS {
+		if commitTS, ok := e.checkTxnCommitted(startTS); ok {
+			return ErrAlreadyCommitted(commitTS)
+		}
+		return nil
+	}
+	e.lock = nil
+	return nil
+}
+
+// MvccStore is an in-memory, multi-versioned, transaction-supported kv storage.
+type MvccStore struct {
+	mu   sync.RWMutex
+	tree *llrb.LLRB
+}
+
+// NewMvccStore creates a MvccStore.
+func NewMvccStore() *MvccStore {
+	return &MvccStore{
+		tree: llrb.New(),
+	}
+}
+
+// Get reads a key by ts.
+func (s *MvccStore) Get(key []byte, startTS uint64) ([]byte, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.get(key, startTS)
+}
+
+func (s *MvccStore) get(key []byte, startTS uint64) ([]byte, error) {
+	entry := s.tree.Get(newEntry(key))
+	if entry == nil {
+		return nil, nil
+	}
+	return entry.(*mvccEntry).Get(startTS)
+}
+
+// A Pair is a KV pair read from MvccStore or an error if any occurs.
+type Pair struct {
+	Key   []byte
+	Value []byte
+	Err   error
+}
+
+// BatchGet gets values with keys and ts.
+func (s *MvccStore) BatchGet(ks [][]byte, startTS uint64) []Pair {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	var pairs []Pair
+	for _, k := range ks {
+		val, err := s.get(k, startTS)
+		if val == nil && err == nil {
+			continue
+		}
+		pairs = append(pairs, Pair{
+			Key:   k,
+			Value: val,
+			Err:   err,
+		})
+	}
+	return pairs
+}
+
+// Scan reads up to limit numbers of Pairs from a key.
+func (s *MvccStore) Scan(startKey, endKey []byte, limit int, startTS uint64) []Pair {
```

Comment:
@shenli Yes, the scan interface may change. But it should always be consistent with TiKV's implementation.